design/20210302.certificate-request-policy.md

```diff
@@ -0,0 +1,509 @@
+---
+title: CertificateRequest Policy
+authors:
+  - "@joshvanl"
+reviewers:
+  - "@cert-manager-dev"
+approvers:
+  - "@cert-manager-dev"
+editor: "@joshvanl"
+creation-date: 2021-03-02
+last-updated: 2021-03-02
+status: provisional
+---
+
+# CertificateRequest Policy
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+- [Goals](#goals)
+- [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  * [API](#api)
+  * [Behaviour](#behaviour)
+  * [RBAC](#rbac)
+- [External Policy Server](#external-policy-server)
+  * [API](#api-1)
+  * [gRPC](#grpc)
+- [User Stories](#user-stories)
+  * [Case 1](#case-1)
+  * [Case 2](#case-2)
+  * [Case 4](#case-4)
+  * [Case 5](#case-5)
+  * [Case 6](#case-6)
+<!-- /toc -->
+
+## Summary
+
+Cluster operators will create and manage a set of policy resources that describe
+what and how certificates can be signed, as well as RBAC to manage who can
+request these certificates. Consumers will be able to request certificates,
+restricted by the policy they are bound to.
+
+
+## Motivation
+
+Policy is an important mechanism for security focused users who would like to
+restrict what, who and how certificates are signed in Kubernetes. Currently, the
+only mechanism for restriction is either via namespacing (via namespace scoped
+Issuer resources) where consumers are bucketed between namespaces, or relying on
+policy that is enforced by the signing Issuer.
+
+This greatly limits the amount of restriction that can be imposed by forcing use
+of certain Issuers, limited protection through namespaing, and no scope for
+users to use external bespoke policy enforcement.
+
+
+## Goals
+- Create first class support for policy over CertificateRequests
+- Allow for Policy to be extended to certificates.k8s.io if cert-manager gains adoption of this API
+- Implement a mechanism for allowing pluggable extensible policy
+- Policy management should have similar user experience to existing Kubernetes
+  resources (network policy, PSP replacement)
+
+## Non-Goals
+- Apply policy to Certificate resources
+- Design external policy engines in this document
```

Comment:
It seems like in this document we _have_ designed a policy engine though. We've defined an API for specifying constraints on what can be specified on a CertificateRequest. I'd expect this document to be focusing on how someone _can_ implement something like what is in this document through the `approval` mechanism.

The `CertificateRequestPolicy` type is in my opinion an *implementation* of a policy engine, not a specification for one.

I'd personally see these objects as effectively being configuration for a *default* policy engine that we have in-built into cert-manager.

Approval in the CSR API is the building block for policy, so us having things like `ExternalPolicyServer` in my opinion makes no sense.

It should be perfectly possible (through the approval API) to implement a policy engine *without* ever installing/creating/touching a 'CertificateRequestPolicy'